packages/sewing-kit-koa/src/manifests.ts

```diff
@@ -0,0 +1,150 @@
+    // least restrictive set of browsers.
+    // 2. If there is only one manifest, use it, regardless of how well it
+    // matches the user agent.
+    // 3. If there is a user agent, find the first manifest where the
+    // browsers it was compiled for matches the user agent, or where there
+    // is no browser restriction on the bundle.
+    // 4. If no matching manifests are found, fall back to the last manifest.
+    //
+    // If a locale-specific match is not found, then the above process is
+    // for repeated for manifests with multi-language support (i.e., no locale
+    // attribute).
+
+    this.resolvedManifestEntry =
+      findMatchingManifest(manifestsByLocale.get(locale!), userAgent) ||
+      findMatchingManifest(multiAsyncLanguageManifests, userAgent) ||
+      multiAsyncLanguageManifests[multiAsyncLanguageManifests.length - 1];
+
+    return this.resolvedManifestEntry;
+  }
+}
+
+let consolidatedManifestPromise: Promise<
+  Map<string, ConsolidatedManifest>
+> | null = null;
+
+function readGzipped(resolvedPath: string) {
+  return readFile(resolvedPath)
+    .then(zippedContent => ungzip(zippedContent))
+    .then(unzippedStr => JSON.parse(unzippedStr.toString()));
+}
+
+function loadConsolidatedManifest(manifestPath: string) {
+  if (consolidatedManifestPromise) {
+    return consolidatedManifestPromise;
+  }
+
+  const resolvedPath = join(appRoot.path, manifestPath);
+  const resolvedZippedPath = `${resolvedPath}.gz`;
+  consolidatedManifestPromise = pathExists(resolvedZippedPath)
+    .then(gzippedVersionExists => {
+      return gzippedVersionExists
+        ? readGzipped(resolvedZippedPath)
+        : readJson(resolvedPath);
+    })
+    .then((manifests: Manifest[]) => {
+      return manifests
+        .map(backfillManifestIdentity)
+        .reduce(
+          groupManifestsByLocale,
+          new Map<string, ConsolidatedManifest>([
+            [MULTI_ASYNC_LANGUAGE_MANIFESTS, []],
+          ]),
+        );
+    });
+
+  return consolidatedManifestPromise;
+}
+
+export function internalOnlyClearCache() {
+  consolidatedManifestPromise = null;
+}
+
+function findMatchingManifest(manifests: Manifest[] | undefined, userAgent) {
+  if (!manifests || manifests.length === 0) {
+    return undefined;
+  }
+
+  if (!userAgent) {
+    return manifests[manifests.length - 1];
+  }
+
+  return (
+    manifests.find(
+      ({browsers}) =>
+        browsers == null ||
+        matchesUA(userAgent, {
+          browsers,
+          ignoreMinor: true,
+          ignorePatch: true,
+          allowHigherVersions: true,
+        }),
+    ) || manifests[manifests.length - 1]
+  );
+}
+
+function backfillManifestIdentity(manifest: Manifest) {
+  if (!manifest.identifier) {
+    manifest.identifier = {target: manifest.name};
+  }
+  return manifest;
+}
+
+function groupManifestsByLocale(
+  acc: Map<string, ConsolidatedManifest>,
+  manifest: Manifest,
+) {
+  const locales = manifest.identifier!.locales;
```

Comment:
is this really optional? If not why don't we change the type (it's from this library isn't it?) if so, we should maybe do an assertion above with a nice error message if it's not there?

Reply:
It's optional for now because `identifier`'s presence depends on an unreleased version of sewing-kit 😅 

The mandatory part can come 🔜 🤞